### PR TITLE
Fix super in async method

### LINKED ISSUE #5124: Async Methods do not support `super`
Given the TypeScript:
```ts
class B extends A {
  async method() { 
    super.superMethod();
  }
}
```

We emit the following JavaScript:
```ts
var __awaiter = ...;
class B extends A {
  method() {
    return __awaiter(this, void 0, Promise, function* () {
      super.superMethod(); // Error
    });
  }
}
```

However, `super` is not bound inside the generator function.

We may instead want to consider changing the emit to the following:

```ts
var __awaiter = ...;
class B extends A {
  method() {
    return __awaiter(this, void 0, Promise, B.prototype._method_async);
  }
  *_method_async() {
    super.superMethod(); // No error
  }
}
```

In this case the `super` binding in `_method_async` works as expected. Also, we pass `B.prototype._method_async` to `__awaiter` here, rather than `this._method_async` so that we will won't need to worry about collisions. Note, though, that we will need to take class expressions into consideration here, as they may not always have a reachable name.

### PR DESCRIPTION
**Pull request by rbuckton**
Supersedes #5141.
Fixes #5124.

With this change, when an async method contains a super call we will create an arrow function that captures the lexical `super` to perform an indexed get (i.e. "geti"), and emit all super property and element access as a call to this function:

```ts
// simple.ts
class A {
  x() { }
}

class B extends A {
  async y() {
    super.x();
  }
}

// simple.js
var __awaiter = ...;
class A {
  x() { }
}

class B extends A {
  y() {
    const _super = name => super[name];
    return __awaiter(this, void 0, Promise, function*() {
      _super("x").call(this);
    });
  }
}
```

However, a function call cannot be the left-hand-side of an assignment (including a destructuring assignment). If such an assignment is detected by the checker, the inline `_super` helper is upgraded to return a wrapper object that can be used on the left-hand side of an assignment:

```ts
// advanced.ts
class A {
  x() { }
}

class B extends A {
  async y() {
    const obj = { x() { } };
    ({ x: super.x } = obj);
  }
}

// advanced.js
var __awaiter = ...;
class A {
  x() { }
}

class B extends A {
  y() {
    const _super = (function (geti, seti) {  
        const cache = Object.create(null);  
        return name => cache[name] || (cache[name] = { get value() { return geti(name); }, set value(v) { seti(name, v); } });  
    })(name => super[name], (name, value) => super[name] = value);
    return __awaiter(this, void 0, Promise, function*() {
      const obj = { x() { } };
      ({ x: _super("x").value } = obj);
    });
  }
}
```

*NOTE: Though we currently attempt to error if you are not accessing a method on super, we don't verify that this is only for a call. As a result, the above sample compiles without error for* `--target ES6`*, and property and element access on* `super` *is legal in ES6 for call, get, and assignment*

### CODE REVIEW

**Comment by rbuckton** [comment]
@mhegazy, @DanielRosenwasser: Can you take a look?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7167,6 +7176,16 @@ namespace ts {
 
             getNodeLinks(node).flags |= nodeCheckFlag;
 
+            // Due to how we emit async functions, we need to specialize the emit for an async method that contains a `super` reference.
+            if (container.kind === SyntaxKind.MethodDeclaration && container.flags & NodeFlags.Async) {
+                if ((isSuperPropertyAccess(node.parent) || isSuperElementAccess(node.parent)) && isAssignmentTarget(node.parent)) {
+                    getNodeLinks(container).flags |= NodeCheckFlags.AsyncMethodWithSuperBinding;
```
Comment (yuit) [inline src/compiler/checker.ts]: Add a comment including example of the case where flags is `NodeCheckFlags.AsyncMethodWithSuperBinding`

File: tests/baselines/reference/asyncImportedPromise_es6.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
         function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
         function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
         function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
-        step((generator = generator.call(thisArg, _arguments)).next());
+        step((generator = generator.apply(thisArg, _arguments)).next());
```
Comment (yuit) [inline tests/baselines/reference/asyncImportedPromise_es6.js]: how is this work before?

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -1488,11 +1488,6 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
             }
 
             function emitExpressionIdentifier(node: Identifier) {
-                if (resolver.getNodeCheckFlags(node) & NodeCheckFlags.LexicalArguments) {
```
Comment (yuit) [inline src/compiler/emitter.ts]: I am slightly confused how this get affected ?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -7167,6 +7176,16 @@ namespace ts {
 
             getNodeLinks(node).flags |= nodeCheckFlag;
 
+            // Due to how we emit async functions, we need to specialize the emit for an async method that contains a `super` reference.
+            if (container.kind === SyntaxKind.MethodDeclaration && container.flags & NodeFlags.Async) {
+                if ((isSuperPropertyAccess(node.parent) || isSuperElementAccess(node.parent)) && isAssignmentTarget(node.parent)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: You could write an `isSuperElementOrPropertyAccess`

File: src/compiler/types.ts [hunk 5]
```diff
@@ -2056,14 +2056,15 @@ namespace ts {
         SuperInstance               = 0x00000100,  // Instance 'super' reference
         SuperStatic                 = 0x00000200,  // Static 'super' reference
         ContextChecked              = 0x00000400,  // Contextual types have been assigned
-        LexicalArguments            = 0x00000800,
-        CaptureArguments            = 0x00001000,  // Lexical 'arguments' used in body (for async functions)
+        AsyncMethodWithSuper        = 0x00000800,
+        AsyncMethodWithSuperBinding = 0x00001000,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: What is the distinction on these? Can you elaborate with comments?

File: tests/baselines/reference/asyncImportedPromise_es6.js (hunk 2 above)
Comment (rbuckton) [inline tests/baselines/reference/asyncImportedPromise_es6.js]: Before, I would rewrite `arguments` to `_arguments` inside of the generator function that serves as the body of the async function. I found it was simpler to remove that rewrite and instead call `apply`. In doing that, the `arguments` object is essentially correct, and the emit is less convoluted.

File: src/compiler/emitter.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: See https://github.com/Microsoft/TypeScript/pull/5870#discussion_r50786034. I found I didn't need this if I switched `__awaiter` from issuing a `generator.call` to a `generator.apply`.

File: src/compiler/types.ts (hunk 5 above)
Comment (rbuckton) [inline src/compiler/types.ts]: I'll add a comment in `checker.ts` per https://github.com/Microsoft/TypeScript/pull/5870#discussion_r50786034

File: src/compiler/checker.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: Fair enough. I updated that in the latest push.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -7167,6 +7172,71 @@ namespace ts {
...
+            //      })(name => super[name], (name, value) => super[name] = value);
+            //      return __awaiter(this, arguments, Promise, function *() {
+            //          [_super("a").value, _super("b").value] = yield ar;
+            //      });
+            //  }
+            //  ...
+            //
+            // This helper creates an object with a "value" property that wraps the `super` property or indexed access for both get and set.
+            // This is required for destructuring assignments, as a call expression cannot be used as the target of a destructuring assignment
+            // while a property access can.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Excellent write-up! :100:

File: src/compiler/emitter.ts [hunk 7]
```diff
@@ -2282,23 +2294,57 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
...
-                    emitSuper(node.expression);
+                let isAsyncMethodWithSuper = false;
+                if (expression.kind === SyntaxKind.SuperKeyword) {
+                    emitSuper(expression);
                     superCall = true;
                 }
                 else {
-                    emit(node.expression);
-                    superCall = node.expression.kind === SyntaxKind.PropertyAccessExpression && (<PropertyAccessExpression>node.expression).expression.kind === SyntaxKind.SuperKeyword;
+                    superCall = isSuperPropertyAccess(expression) || isSuperElementAccess(expression);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Use the new helper here

File: src/compiler/types.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: A quick comment would still be nice. Here's something for now:

```ts
        AsyncMethodWithSuper        = 0x00000800, // An async method that refers to to 'super' without assigning to its members.
        AsyncMethodWithSuperBinding = 0x00001000, // An async method that refers to 'super' where members are assigned to.
```

**Comment by yuit** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**